LoaderCallbacks.onLoaderReset & NPE

I’m resetting cursor at `LoaderCallbacks#onLoaderReset` as [described at the documentation](http://developer.android.com/guide/components/loaders.html#callback). Unfortunately this action fails for `IterableCursorAdapter`. Trace is below for `cursor-utils:0.3`.

Actually I’m not sure why it happens because `instance of` check [should work just fine with `null` values](http://stackoverflow.com/a/2950415/3359826).

```
java.lang.RuntimeException: Unable to destroy activity: java.lang.NullPointerException
  at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3273)
  at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:3291)
  at android.app.ActivityThread.access$1200(ActivityThread.java:130)
  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1248)
  at android.os.Handler.dispatchMessage(Handler.java:99)
  at android.os.Looper.loop(Looper.java:137)
  at android.app.ActivityThread.main(ActivityThread.java:4745)
  at java.lang.reflect.Method.invokeNative(Native Method)
  at java.lang.reflect.Method.invoke(Method.java:511)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
  at dalvik.system.NativeStart.main(Native Method)
Caused by: java.lang.NullPointerException
  at com.venmo.cursor.support.IterableCursorAdapter.enforceIterableCursor(IterableCursorAdapter.java:62)
  at com.venmo.cursor.support.IterableCursorAdapter.swapCursor(IterableCursorAdapter.java:44)
  at application.id.fragment.ListFragment.onLoaderReset(ListFragment.java:74)
  at android.support.v4.app.LoaderManagerImpl$LoaderInfo.destroy(LoaderManager.java:339)
  at android.support.v4.app.LoaderManagerImpl.doDestroy(LoaderManager.java:776)
  at android.support.v4.app.Fragment.onDestroy(Fragment.java:1202)
  at android.support.v4.app.Fragment.performDestroy(Fragment.java:2006)
  at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1076)
  at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1126)
  at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1108)
  at android.support.v4.app.FragmentManagerImpl.dispatchDestroy(FragmentManager.java:1954)
  at android.support.v4.app.FragmentActivity.onDestroy(FragmentActivity.java:313)
  at android.support.v7.app.ActionBarActivity.onDestroy(ActionBarActivity.java:169)
  at android.app.Activity.performDestroy(Activity.java:5172)
  at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1109)
  at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3260)
  at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:3291)
  at android.app.ActivityThread.access$1200(ActivityThread.java:130)
  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1248)
  at android.os.Handler.dispatchMessage(Handler.java:99)
  at android.os.Looper.loop(Looper.java:137)
  at android.app.ActivityThread.main(ActivityThread.java:4745)
  at java.lang.reflect.Method.invokeNative(Native Method)
  at java.lang.reflect.Method.invoke(Method.java:511)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
  at dalvik.system.NativeStart.main(Native Method)
```

--- a/cursor-utils/src/androidTest/java/com/venmo/cursor/IterableCursorAdapterTest.java
+++ b/cursor-utils/src/androidTest/java/com/venmo/cursor/IterableCursorAdapterTest.java
@@ -1,0 +1,82 @@
+package com.venmo.cursor;
+
+import android.content.Context;
+import android.test.AndroidTestCase;
+import android.view.View;
+import android.view.ViewGroup;
+
+import java.util.Arrays;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class IterableCursorAdapterTest extends AndroidTestCase {
+
+    private class TestAdapter extends IterableCursorAdapter<Object> {
+
+        private TestAdapter(IterableCursor<Object> c) {
+            super(getContext(), c, false /* autoRequery */);
+        }
+
+        @Override
+        public View newView(Context context, Object o, ViewGroup parent) {
+            return new View(context);
+        }
+
+        @Override
+        public void bindView(View view, Context context, Object o) {
+            // intentionally empty
+        }
+    }
+
+    public void testCanSwapWithNullCursor() {
+        TestAdapter adapter = new TestAdapter(new CursorList<>());
+        adapter.swapCursor(null);
+    }
+
+    public void testCanReturnNullCursorFromQuery() throws InterruptedException {
+        final AtomicBoolean wasCalled = new AtomicBoolean(false);
+        final CountDownLatch latch = new CountDownLatch(1);
+        TestAdapter adapter = new TestAdapter(new CursorList<>()) {
+            @Override
+            public IterableCursor<Object> runQueryOnBackgroundThread(CharSequence constraint) {
+                try {
+                    return null;
+                } finally {
+                    wasCalled.set(true);
+                    latch.countDown();
+                }
+            }
+        };
+        adapter.getFilter().filter(null);
+        latch.await(1, TimeUnit.SECONDS);
+        assertTrue(wasCalled.get());
+    }
+
+    public void testCanUseNullCursorForConstructor() {
+        new TestAdapter(null);
+    }
+
+    public void testBindViewGetsCorrectObject() {
+        Object obj1 = "obj1";
+        Object obj2 = "obj2";
+        final AtomicReference<Object> nextObject = new AtomicReference<>();
+        final AtomicInteger counter = new AtomicInteger();
+        TestAdapter adapter = new TestAdapter(new CursorList<>(Arrays.asList(obj1, obj2))) {
+            @Override
+            public void bindView(View view, Context context, Object o) {
+                assertEquals(o, nextObject.get());
+                counter.incrementAndGet();
+                super.bindView(view, context, o);
+            }
+        };
+        nextObject.set(obj1);
+        adapter.bindView(new View(getContext()), getContext(), adapter.getCursor());
+        adapter.getCursor().moveToNext(); // simulate list scrolling
+        nextObject.set(obj2);
+        adapter.bindView(new View(getContext()), getContext(), adapter.getCursor());
+        assertEquals(2, counter.get());
+    }
+}

--- a/cursor-utils/src/androidTest/java/com/venmo/cursor/support/SupportIterableCursorAdapterTest.java
+++ b/cursor-utils/src/androidTest/java/com/venmo/cursor/support/SupportIterableCursorAdapterTest.java
@@ -1,0 +1,85 @@
+package com.venmo.cursor.support;
+
+import android.content.Context;
+import android.test.AndroidTestCase;
+import android.view.View;
+import android.view.ViewGroup;
+
+import com.venmo.cursor.CursorList;
+import com.venmo.cursor.IterableCursor;
+
+import java.util.Arrays;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class SupportIterableCursorAdapterTest extends AndroidTestCase {
+
+    private class TestAdapter extends IterableCursorAdapter<Object> {
+
+        private TestAdapter(IterableCursor<Object> c) {
+            super(getContext(), c, false /* autoRequery */);
+        }
+
+        @Override
+        public View newView(Context context, Object o, ViewGroup parent) {
+            return new View(context);
+        }
+
+        @Override
+        public void bindView(View view, Context context, Object o) {
+            // intentionally empty
+        }
+    }
+
+    public void testCanSwapWithNullCursor() {
+        TestAdapter adapter = new TestAdapter(new CursorList<>());
+        adapter.swapCursor(null);
+    }
+
+    public void testCanReturnNullCursorFromQuery() throws InterruptedException {
+        final AtomicBoolean wasCalled = new AtomicBoolean(false);
+        final CountDownLatch latch = new CountDownLatch(1);
+        TestAdapter adapter = new TestAdapter(new CursorList<>()) {
+            @Override
+            public IterableCursor<Object> runQueryOnBackgroundThread(CharSequence constraint) {
+                try {
+                    return null;
+                } finally {
+                    wasCalled.set(true);
+                    latch.countDown();
+                }
+            }
+        };
+        adapter.getFilter().filter(null);
+        latch.await(1, TimeUnit.SECONDS);
+        assertTrue(wasCalled.get());
+    }
+
+    public void testCanUseNullCursorForConstructor() {
+        new TestAdapter(null);
+    }
+
+    public void testBindViewGetsCorrectObject() {
+        Object obj1 = "obj1";
+        Object obj2 = "obj2";
+        final AtomicReference<Object> nextObject = new AtomicReference<>();
+        final AtomicInteger counter = new AtomicInteger();
+        TestAdapter adapter = new TestAdapter(new CursorList<>(Arrays.asList(obj1, obj2))) {
+            @Override
+            public void bindView(View view, Context context, Object o) {
+                assertEquals(o, nextObject.get());
+                counter.incrementAndGet();
+                super.bindView(view, context, o);
+            }
+        };
+        nextObject.set(obj1);
+        adapter.bindView(new View(getContext()), getContext(), adapter.getCursor());
+        adapter.getCursor().moveToNext(); // simulate list scrolling
+        nextObject.set(obj2);
+        adapter.bindView(new View(getContext()), getContext(), adapter.getCursor());
+        assertEquals(2, counter.get());
+    }
+}

--- a/cursor-utils/src/main/java/com/venmo/cursor/IterableCursorAdapter.java
+++ b/cursor-utils/src/main/java/com/venmo/cursor/IterableCursorAdapter.java
@@ -27,7 +27,7 @@ public abstract class IterableCursorAdapter<T> extends CursorAdapter {
     public abstract View newView(Context context, T t, ViewGroup parent);
 
     @Override
-    public void bindView(View view, Context context, Cursor cursor) {
+    public final void bindView(View view, Context context, Cursor cursor) {
         T t = getCursor().peek();
         bindView(view, context, t);
     }
@@ -59,6 +59,7 @@ public abstract class IterableCursorAdapter<T> extends CursorAdapter {
     }
 
     private IterableCursor<T> enforceIterableCursor(Cursor cursor) {
+        if (cursor == null) return null;
         if (!(cursor instanceof IterableCursor)) {
             throw new IllegalArgumentException(
                     cursor.getClass().getName() + " is not an " + IterableCursor.class.getName());

--- a/cursor-utils/src/main/java/com/venmo/cursor/support/IterableCursorAdapter.java
+++ b/cursor-utils/src/main/java/com/venmo/cursor/support/IterableCursorAdapter.java
@@ -26,7 +26,7 @@ public abstract class IterableCursorAdapter<T> extends CursorAdapter {
     public abstract View newView(Context context, T t, ViewGroup parent);
 
     @Override
-    public void bindView(View view, Context context, Cursor cursor) {
+    public final void bindView(View view, Context context, Cursor cursor) {
         T t = getCursor().peek();
         bindView(view, context, t);
     }
@@ -58,6 +58,7 @@ public abstract class IterableCursorAdapter<T> extends CursorAdapter {
     }
 
     private IterableCursor<T> enforceIterableCursor(Cursor cursor) {
+        if (cursor == null) return null;
         if (!(cursor instanceof IterableCursor)) {
             throw new IllegalArgumentException(
                     cursor.getClass().getName() + " is not an " + IterableCursor.class.getName());

